packages/theming/src/utils/arrowStyles.ts

```diff
@@ -150,45 +120,47 @@ const positionStyles = (position: ArrowPosition, size: string, inset: string) =>
  * @component
  */
 export default function arrowStyles(position: ArrowPosition, options: ArrowOptions = {}) {
-  const size = options.size || '6px';
   const inset = options.inset || '0';
-  const squareSize = math(`${size} * 2 / sqrt(2)`, exponentialSymbols);
+  const size = options.size === undefined ? 6 : (stripUnit(options.size) as number);
+  const squareSize = `${Math.round((size * 2) / Math.sqrt(2))}px`;
+  const afterOffset = 0;
+  const beforeOffset = afterOffset + 2;
 
   /**
    * 1. Set base positioning for an element with an arrow.
-   * 2. Allow any border inherited by `::after` to show through.
-   * 3. Border styling and box-shadow will be automatically inherited from the
-   *    parent element.
-   * 4. Apply shared sizing properties to ::before and ::after.
+   * 2. Apply shared properties to ::before and ::after.
+   * 3. Display border with inherited border-color
+   * 4. Clip the inner square forming the arrow body into a triangle so that it
+   *    doesn't interfere with container content.
+   * 5. Clip the outer square forming the arrow border into a triangle so that the
+   *    border merge with the container's.
    */
   return css`
     position: relative; /* [1] */
 
-    &::before {
+    &::before,
+    &::after {
       /* [2] */
+      position: absolute;
       border-width: inherit;
       border-style: inherit;
-      border-color: transparent;
-      background-clip: content-box;
+      width: ${squareSize};
+      height: ${squareSize};
+      content: '';
+      box-sizing: inherit;
     }
 
-    &::after {
-      /* [3] */
-      z-index: -1;
-      border: inherit;
-      box-shadow: inherit;
```

Comment:
Can `box-shadow` inheritance be restored?

Reply:
Unfortunately, restoring `box-shadow` will not work. Using a clipping mask on both layers will clip the drop shadow. 😞